src/tools/rbd/action/Group.cc

```diff
@@ -745,124 +769,1532 @@ int execute_group_snap_rollback(const po::variables_map &vm,
+{
+  vector<librbd::group_snap_info_t> snaps;
+
+  int r = rbd.group_snap_list(group_ioctx, group_name, &snaps,
+                              sizeof(librbd::group_snap_info_t));
+  if (r < 0) {
+    return r;
+  }
+
+  is_exist = false;
+  for (auto snap : snaps) {
+    if (snap.name == group_snap_name) {
+      is_exist = true;
+      break;
+    }
+  }
+
+  return r;
+}
+
+static int do_group_snap_from(librbd::RBD &rbd, librados::IoCtx& group_ioctx,
+                              const char *group_name, int fd)
+{
+  int r;
+  string from;
+
+  // 4k limit to make sure we don't get a garbage string
+  r = group_read_string(fd, 4096, &from);
+  if (r < 0) {
+    std::cerr << "rbd: failed to decode start snap name" << std::endl;
+    return r;
+  }
+
+  bool exist = false;
+  r = group_snap_is_exist(rbd, group_ioctx, group_name, from, exist);
+  if (r < 0) {
+    return r;
+  }
+
+  if (!exist) {
+    std::cerr << "start snapshot '" << from
+              << "' does not exist in the image, aborting" << std::endl;
+    return -EINVAL;
+  }
+
+  return r;
+}
+
+static int do_group_image_snap_to(librbd::RBD &rbd, librados::IoCtx& group_ioctx,
+                                  const char *group_name,
+                                  int fd, std::string &tosnap)
+{
+  int r;
+  string to;
+  // 4k limit to make sure we don't get a garbage string
+  r = group_read_string(fd, 4096, &to);
+  if (r < 0) {
+    std::cerr << "rbd: failed to decode end snap name" << std::endl;
+    return r;
+  }
+
+  bool exists;
+  r = group_snap_is_exist(rbd, group_ioctx, group_name, to, exists);
+  if (r < 0) {
+    return r;
+  }
+
+  if (exists) {
+    std::cerr << "end snapshot '" << to << "' already exists, aborting" << std::endl;
+    return -EEXIST;
+  }
+
+  tosnap = to;
+
+  return r;
+}
+
+static int group_import_snap_info(librbd::RBD &rbd, librados::IoCtx& group_ioctx,
+                        const char *group_name, int fd,
+                        string &to_snap, const int import_format)
+{
+  int r = utils::validate_banner(fd, utils::RBD_GROUP_BANNER_SNAP);
+  if (r < 0) {
+    return r;
+  }
+
+  while (r == 0) {
+    __u8 tag;
+    uint64_t length = 0;
+    r = utils::read_tag(fd, RBD_EXPORT_IMAGE_END, import_format, &tag, &length);
+    if (r < 0 || tag == RBD_EXPORT_IMAGE_END) {
+      break;
+    }
+
+    if (RBD_DIFF_FROM_SNAP == tag) {
+      r = do_group_snap_from(rbd, group_ioctx, group_name, fd);
+    } else if (RBD_DIFF_TO_SNAP == tag) {
+      r = do_group_image_snap_to(rbd, group_ioctx, group_name, fd, to_snap);
+    } else {
+      std::cerr << "rbd: invalid tag in image properties zone: " << tag
+                << ", Skip it." << std::endl;
+      r = utils::skip_tag(fd, length);
+    }
+  }
+
+  return r;
+}
+
+static int group_import_image_info(int fd,
+                        std::string &src_pool_name, std::string &src_image_name,
+                        uint64_t &is_create_image, const int import_format)
```

Comment:
nit: the convention is to use pointers (not references) for the output parameters. And place the input parameters first.